tests/unit_tests/components/config_parser.cpp

```diff
@@ -0,0 +1,247 @@
+  EXPECT_EQ(GetParam().first.first, line.key);
+  EXPECT_EQ(GetParam().first.second, line.value);
+}
+
+TEST_F(ParseLineInValidTest, throwsSyntaxError) {
+  EXPECT_THROW(parser->parse_line("unknown"), syntax_error);
+}
+// }}}
+
+// GetLineTypeTest {{{
+
+/**
+ * \brief Class for parameterized tests on get_line_type
+ *
+ * Parameters are pairs of the expected line type and a string that should be
+ * detected as that line type
+ */
+class GetLineTypeTest :
+  public ConfigParser,
+  public ::testing::WithParamInterface<pair<line_type, string>> {
+  };
+
+/**
+ * \brief Helper function generate GetLineTypeTest parameter values
+ */
+vector<pair<line_type, string>> line_type_transform(vector<string>&& in, line_type type) {
+  vector<pair<line_type, string>> out;
+
+  out.reserve(in.size());
+  for (const auto& i : in) {
+    out.emplace_back(type, i);
+  }
+
+  return out;
+}
+
+
+/**
+ * \brief Parameter values for GetLineTypeTest
+ */
+auto line_type_key = line_type_transform({"a = b", "  a =b", " a\t =\t \t b", "a = "}, line_type::KEY);
+auto line_type_header = line_type_transform({"[section]", "[section]", "[section/sub]"}, line_type::HEADER);
+auto line_type_comment = line_type_transform({";abc", "#abc", ";", "#"}, line_type::COMMENT);
+auto line_type_empty = line_type_transform({""}, line_type::EMPTY);
+auto line_type_unknown = line_type_transform({"|a", " |a", "a"}, line_type::UNKNOWN);
+
+/**
+ * Instantiate GetLineTypeTest for the different line types
+ */
+INSTANTIATE_TEST_SUITE_P(LineTypeKey, GetLineTypeTest, ::testing::ValuesIn(line_type_key));
+INSTANTIATE_TEST_SUITE_P(LineTypeHeader, GetLineTypeTest, ::testing::ValuesIn(line_type_header));
+INSTANTIATE_TEST_SUITE_P(LineTypeComment, GetLineTypeTest, ::testing::ValuesIn(line_type_comment));
+INSTANTIATE_TEST_SUITE_P(LineTypeEmpty, GetLineTypeTest, ::testing::ValuesIn(line_type_empty));
+INSTANTIATE_TEST_SUITE_P(LineTypeUnknown, GetLineTypeTest, ::testing::ValuesIn(line_type_unknown));
+
+/**
+ * \brief Parameterized test for get_line_type
+ */
+TEST_P(GetLineTypeTest, correctness) {
+  EXPECT_EQ(GetParam().first, parser->get_line_type(GetParam().second));
+}
+
+// }}}
+
+// ParseKeyTest {{{
+
+/**
+ * \brief Class for parameterized tests on parse_key
+ *
+ * The first element of the pair is the expected key-value pair and the second
+ * element is the string to be parsed, has to be trimmed and valid.
+ */
+class ParseKeyTest :
+  public ConfigParser,
+  public ::testing::WithParamInterface<pair<pair<string, string>, string>> {};
+
+
+vector<pair<pair<string, string>, string>> parse_key_list = {
+  {{"key", "value"}, "key = value"},
+  {{"key", "value"}, "key=value"},
+  {{"key", "value"}, "key =\"value\""},
+  {{"key", "value"}, "key\t=\t \"value\""},
+  {{"key", "\"value"}, "key = \"value"},
+  {{"key", "value\""}, "key = value\""},
+  {{"key", "= value"}, "key == value"},
+  {{"key", ""}, "key ="},
+  {{"key", ""}, "key =\"\""},
+  {{"key", "\"\""}, "key =\"\"\"\""},
```

Comment:
```suggestion
  {{"key", "\"\""}, R"(key ="""")"},
```
This isn't mandatory, but just a comment in case you don't know this feature.
It can be more readable and less error prone.
See https://en.cppreference.com/w/cpp/language/string_literal for more information.

Reply:
Oh, I didn't know about this. That's great. Too bad we can't use it in strings where we use `\t`.

Reply:
Yeah that's a new feature of C++14, especially useful when you are writing regexes.